File: Link/DLL/explicitLink.cpp
// Using explicit run-time linking
#include <windows.h>
#include <stdio.h>

// DLL function signature
typedef double (*importFunction)(double,double);

int main(int argc,char**argv)
{
    importFunction addNumbers;
    double result;
    
    // Load DLL file
    HINSTANCE hinstLib = LoadLibrary(TEXT("Example.dll"));
    if (hinstLib == NULL)
    {
        printf("ERROR: unable to load DLL\n");
        return 1;
    }
    
    // Get function pointer
    addNumbers = (importFunction) GetProcAddress(hinstLib,"AddNumbers");
    if (addNumbers == NULL)
    {
        printf("ERROR: unable to find DLL function\n");
        FreeLibrary(hinstLib);
        return 2;
    }
    
    // Call function.
    result = addNumbers(1,2);
    
    // Unload DLL file
    FreeLibrary(hinstLib);
    
    // Display result
    printf("The result was : %f\n", result);
    
    return 0;
}

File: Link/DLL/explicitLink.pas
program ExplicitLink;
{Using Explicit Run-Time Linking}

{$APPTYPE CONSOLE}

uses Windows;

var
   AddNumbers : function (a,b:Double):Double; cdecl;
   LibHandle  : HMODULE;

begin
   LibHandle := LoadLibrary('example.dll');
   
   if LibHandle = 0 then Exit;
   
   AddNumbers := GetProcAddress(LibHandle, 'AddNumbers');
   
   if Assigned(AddNumbers) then
      WriteLn('1 + 2 = ', AddNumbers(1,2))
   else
      WriteLn('Error: unable to find DLL function');
   
   FreeLibrary(LibHandle);
end.
